Improve Pythonic file interface for LittleFS.open()

cesarvandevelde opened the pull request:
This pull request improves the Pythonic API by reusing the base classes from the builtin `io` module. These changes make littlefs file objects behave similar to standard Python file-like objects:

- Files can be opened in either text or binary mode, and will return `str` or
  `bytes` respectively.
- `read()` will now read until EOF unless length is specified.
- `open()` now supports all standard mode flags, e.g. 'a' for append, 'w+' for
  read/write, etc.
- `open()` raises standard exceptions (e.g. `FileNotFoundError`) where applicable.

jrast commented:
Thank you for this PR. I will look into this in the next weeks. Please ping me if I forget about it!

cesarvandevelde commented:
Great, thanks!

As an aside: I noticed that CI fails on Python 3.5 due to an f-string. The fix is simple, but maybe it's better for the library to use the same supported versions as CPython (3.5 and 3.6 are end-of-life).

jrast commented on src/littlefs/lfs.pyx:
```diff
@@ -186,13 +201,58 @@ def removeattr(LFSFilesystem fs, path, type):
     raise NotImplementedError
 
 
+def _parse_mode(mode) -> LFSFileFlag:
```
Would it be possible to consolidate the parsing code which is used here with the code used in LittleFS.open() ?
At least the string -> boolean flags conversion seems (should) be the same.

jrast reviewed (commented):
Nice work and a great improvement for the library! I still need to play around with the library a little bit (long not used...) to get a feeling about the extended interface but it looks great.

jrast commented:
> Great, thanks!
> 
> As an aside: I noticed that CI fails on Python 3.5 due to an f-string. The fix is simple, but maybe it's better for the library to use the same supported versions as CPython (3.5 and 3.6 are end-of-life).

Sure, Python 3.5 can be dropped. Python 3.6 is still in use on some of our devices, so i prefer to keep this.

cesarvandevelde commented:
I've moved the flag parsing logic into the `file_open` method. Initially, I thought that I would reuse the parsing logic in the `LittleFS.open()` method, hence the separate helper function. However, these two methods are sufficiently distinct, to the point where code sharing does not make sense.

I've also updated the docs to reflect the changes of this PR:

- Added / removed the `b` prefix in the code examples
- Added docstrings to new methods

> Sure, Python 3.5 can be dropped. Python 3.6 is still in use on some of our devices, so i prefer to keep this.

Sounds good to me! I'm not familiar with Appveyor, but maybe you could add Python 3.9 and 3.10 to the environment matrix?

jrast commented:
Thank you again for your additions! I merged them into master. However, Travis-CI which was used to build the wheels is no longer free so I need to invest some time to move to Github Actions in order to create a new release.

jrast commented:
Ok, I made it! A new release is available on pypi:
https://pypi.org/project/littlefs-python/0.3.0/

cesarvandevelde commented:
Happy to help! I'm currently working on a project that uses an embedded LFS file system, and this library has been an excellent tool for preparing and inspecting disk images. So thank you for this excellent library!
